yt/frontends/gadget/io.py

```diff
@@ -77,13 +80,58 @@ def _yield_coordinates(self, data_file, needed_ptype=None):
             yield key, pos
         f.close()
 
+    def _generate_smoothing_length(self, data_files, kdtree):
```

Comment:
The logic in this routine is a bit squirrely and I would like @matthewturk and others to check it.